File: interface/src/apps/marketplace/listing-status.ts
/**
 * Discoverability setting for an agent. `closed` keeps the agent private to
 * its creator (and their org); `hireable` lists it in the Marketplace so
 * other users can hire it into their projects.
 *
 * Phase 1 encodes this in the agent's `tags` via `LISTING_STATUS_TAG_PREFIX`;
 * Phase 3 promotes it to a dedicated `listing_status` column on the network
 * agent record.
 */
export type AgentListingStatus = "closed" | "hireable";

export const DEFAULT_LISTING_STATUS: AgentListingStatus = "closed";

export const LISTING_STATUS_TAG_PREFIX = "listing_status:";

export function listingStatusFromTags(
  tags: readonly string[] | undefined,
): AgentListingStatus {
  if (!tags || tags.length === 0) return DEFAULT_LISTING_STATUS;
  for (const tag of tags) {
    if (tag.startsWith(LISTING_STATUS_TAG_PREFIX)) {
      const value = tag.slice(LISTING_STATUS_TAG_PREFIX.length).toLowerCase();
      if (value === "hireable" || value === "closed") {
        return value;
      }
    }
  }
  return DEFAULT_LISTING_STATUS;
}

export function mergeListingStatusTag(
  existing: readonly string[] | undefined,
  listingStatus: AgentListingStatus,
): string[] {
  const kept = (existing ?? []).filter(
    (t) => !t.toLowerCase().startsWith(LISTING_STATUS_TAG_PREFIX),
  );
  if (listingStatus !== DEFAULT_LISTING_STATUS) {
    kept.push(`${LISTING_STATUS_TAG_PREFIX}${listingStatus}`);
  }
  return kept;
}

File: crates/aura-os-core/src/listing_status.rs
//! Marketplace listing status for agents.
//!
//! Phase 1 and Phase 2 encoded this as a `listing_status:<value>` tag on
//! `Agent.tags`. Phase 3 promotes it to a typed field on `Agent`. The tag
//! form is still read as a fallback on the frontend loader, but the server
//! and network client treat the typed field as the source of truth.

use serde::{Deserialize, Serialize};

/// Whether an agent is publicly listed for hire on the marketplace.
///
/// Defaults to [`AgentListingStatus::Closed`] so existing agents without a
/// stored value stay private until their owner opts in.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum AgentListingStatus {
    /// The agent is not listed on the marketplace.
    #[default]
    Closed,
    /// The agent is available for hire on the marketplace.
    Hireable,
}

/// Tag prefix used by Phase 1/Phase 2 to encode this value on `Agent.tags`.
/// Retained so the frontend loader can fall back to tags for agents that
/// have not yet been backfilled with typed fields.
pub const LISTING_STATUS_TAG_PREFIX: &str = "listing_status:";

impl AgentListingStatus {
    /// Parse a marketplace listing status from a string. Accepts
    /// `"closed"` and `"hireable"` case-insensitively, surrounding
    /// whitespace is trimmed.
    pub fn from_str(value: &str) -> Result<Self, String> {
        match value.trim().to_ascii_lowercase().as_str() {
            "closed" => Ok(Self::Closed),
            "hireable" => Ok(Self::Hireable),
            other => Err(format!("unknown listing_status: {other}")),
        }
    }

    /// Canonical lowercase string representation. Matches the serde form.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Closed => "closed",
            Self::Hireable => "hireable",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_is_closed() {
        assert_eq!(AgentListingStatus::default(), AgentListingStatus::Closed);
    }

    #[test]
    fn from_str_accepts_canonical_lowercase() {
        assert_eq!(
            AgentListingStatus::from_str("closed").unwrap(),
            AgentListingStatus::Closed
        );
        assert_eq!(
            AgentListingStatus::from_str("hireable").unwrap(),
            AgentListingStatus::Hireable
        );
    }

    #[test]
    fn from_str_is_case_insensitive_and_trims_whitespace() {
        assert_eq!(
            AgentListingStatus::from_str("  Closed  ").unwrap(),
            AgentListingStatus::Closed
        );
        assert_eq!(
            AgentListingStatus::from_str("HIREABLE").unwrap(),
            AgentListingStatus::Hireable
        );
    }

    #[test]
    fn from_str_rejects_unknown_values() {
        let err = AgentListingStatus::from_str("pending").expect_err("unknown value must fail");
        assert!(err.contains("pending"), "error was: {err}");
    }

    #[test]
    fn as_str_roundtrips_through_from_str() {
        for status in [AgentListingStatus::Closed, AgentListingStatus::Hireable] {
            assert_eq!(
                AgentListingStatus::from_str(status.as_str()).unwrap(),
                status
            );
        }
    }

    #[test]
    fn serde_uses_snake_case() {
        let json = serde_json::to_string(&AgentListingStatus::Hireable).unwrap();
        assert_eq!(json, "\"hireable\"");
        let parsed: AgentListingStatus = serde_json::from_str("\"closed\"").unwrap();
        assert_eq!(parsed, AgentListingStatus::Closed);
    }
}
